### PR TITLE
 [feature-wip](parquet-vec) Support parquet scanner in vectorized engine

### PR DESCRIPTION
**Pull request by yinzhijian**
# Proposed changes

Issue Number: close #xxx

## Optimization Todo list：
1. FE generates the corresponding src slot desc and Expr through the parquet schema
2. BE supports direct conversion of arrow type into dest primitive type of similar type. For example, arrow type is INT32, and dest type is TYPE_BIGINT (int64), INT32=>TYPE_BIGINT. Instead of the current way: INT32=> TYPE_INT => TYPE_BIGINT

## Performance Testing:
load parquet file in vec version almost 1x faster than rowset version.
rows num:300k
test table schema:
CREATE TABLE `parquet` (
  `id` int(11) NOT NULL COMMENT "",
  `email` varchar(26) NOT NULL COMMENT "",
  `c_date32` DATE NOT NULL COMMENT "",
  `c_date64` DATETIME NOT NULL COMMENT "",
  `c_timestamp` DATETIME NOT NULL COMMENT "",
  `c_decimal128` DECIMAL(27, 9) NULL COMMENT "",
  `c_bool` BOOLEAN NULL COMMENT "",
  `c_float` FLOAT NULL COMMENT "",
  `c_double` DOUBLE NULL COMMENT "",
  `c_fixed_size_binary` CHAR(20) NULL COMMENT "",
  `c_binary` VARCHAR(32) NULL COMMENT "",
  `c_uint64` BIGINT NULL COMMENT ""
)
DISTRIBUTED BY HASH(`id`) BUCKETS 1
PROPERTIES (
"replication_num" = "1"
);

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
4. Has document been added or modified: (No Need)
5. Does it need to update dependencies: (No)
6. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/utils/arrow_column_to_doris_column.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "common/status.h"
+#include "runtime/primitive_type.h"
+#include "vec/core/column_with_type_and_name.h"
+
+// This files contains some utilities to convert Doris internal
+// data format from Apache Arrow format. 
+namespace doris::vectorized {
+
+const PrimitiveType arrow_type_to_pt(::arrow::Type::type type);
```
Comment (yangzhg) [inline be/src/vec/utils/arrow_column_to_doris_column.h:33]: `pt` feels a little weird.  `primitive_type` maybe better

File: be/src/vec/utils/arrow_column_to_doris_column.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+Status arrow_column_to_doris_column(const arrow::Array* arrow_column,
+                                    size_t arrow_batch_cur_idx,
+                                    ColumnWithTypeAndName& doirs_column,
+                                    size_t num_elements,
+                                    const std::string& timezone);
```
Comment (yangzhg) [inline be/src/vec/utils/arrow_column_to_doris_column.h:39]: why all types need timezone parmater

File: be/src/http/action/stream_load.cpp [hunk 3]
```diff
@@ -98,6 +98,8 @@ static TFileFormatType::type parse_format(const std::string& format_str,
         if (compress_type.empty()) {
             format_type = TFileFormatType::FORMAT_JSON;
         }
+    } else if (boost::iequals(format_str, "PARQUET")) {
```
Comment (yangzhg) [inline be/src/http/action/stream_load.cpp:101]: Better to use  `iequals` in `be/src/util/string_util.h`,  avoid using functions from boost

File: be/src/vec/exec/vparquet_scanner.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    }
+    return Status::EndOfFile("EOF");
+}
+
+Status VParquetScanner::init_arrow_batch_if_necessary() {
+    // 1. init batch if first time 
+    // 2. reset reader if end of file
+    if (_scanner_eof || _batch == nullptr || _arrow_batch_cur_idx >= _batch->num_rows()) {
+        while (!_scanner_eof) {
+            Status status = next_arrow_batch();
```
Comment (yangzhg) [inline be/src/vec/exec/vparquet_scanner.cpp:92]: put `Status status` outside of while

File: be/src/vec/exec/vparquet_scanner.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    // range of current file
+    fill_columns_from_path(&src_block);
+    RETURN_IF_ERROR(eval_conjunts(&src_block));
+    // materialize, src block => dest columns
+    RETURN_IF_ERROR(materialize_block(&src_block, columns));
+    if (_scanner_eof) {
+        *eof = true;
+    } else {
+        *eof = false;
+    }
```
Comment (yangzhg) [inline be/src/vec/exec/vparquet_scanner.cpp:186]: *eof = _scanner_eof

File: be/src/vec/utils/arrow_column_to_doris_column.h (hunk 2 above)
Comment (yinzhijian) [reply be/src/vec/utils/arrow_column_to_doris_column.h:39]: there is no better way than this

File: be/src/vec/CMakeLists.txt [hunk 6]
```diff
@@ -189,6 +190,11 @@ set(VEC_FILES
   runtime/vdata_stream_recvr.cpp
   runtime/vdata_stream_mgr.cpp
   runtime/vpartition_info.cpp
+  runtime/vsorted_run_merger.cpp
+  runtime/vload_channel.cpp
```
Comment (HappenLee) [inline be/src/vec/CMakeLists.txt:194]: the code not add in this PR？

File: be/src/vec/utils/arrow_column_to_doris_column.cpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+#include "arrow/array/array_binary.h"
+#include "arrow/array/array_nested.h"
+#include "arrow/scalar.h"
+#include "arrow/type_fwd.h"
+#include "arrow/type_traits.h"
+
+#define FOR_ARROW_TYPES(M) \
+    M(::arrow::Type::BOOL, TYPE_BOOLEAN) \
+    M(::arrow::Type::INT8, TYPE_TINYINT) \
+    M(::arrow::Type::UINT8, TYPE_TINYINT) \
```
Comment (HappenLee) [inline be/src/vec/utils/arrow_column_to_doris_column.cpp:38]: UINT8 bigger than tinyint ？may overflow

File: be/src/vec/utils/arrow_column_to_doris_column.cpp [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+/// Inserts chars and offsets right into internal column data to reduce an overhead.
+/// Internal offsets are shifted by one to the right in comparison with Arrow ones. So the last offset should map to the end of all chars.
+/// Also internal strings are null terminated.
+static Status convert_column_with_string_data(const arrow::Array* array, size_t array_idx,
+                                        MutableColumnPtr& data_column, size_t num_elements) {
+    PaddedPODArray<UInt8> & column_chars_t = assert_cast<ColumnString &>(*data_column).get_chars();
+    PaddedPODArray<UInt32> & column_offsets = assert_cast<ColumnString &>(*data_column).get_offsets();
+
+    const auto & concrete_array = dynamic_cast<const arrow::BinaryArray &>(*array);
```
Comment (HappenLee) [inline be/src/vec/utils/arrow_column_to_doris_column.cpp:102]: code format： auto&. same to other.

Comment (HappenLee) [inline be/src/vec/utils/arrow_column_to_doris_column.cpp:102]: Rethink here really need `dynamic_cast` ?
1. `dynamic_cast` means bad design of class or code, replace with virtual function
2.  this code may throw exception if cast faild. it is danger here.

File: be/src/vec/utils/arrow_column_to_doris_column.cpp [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+            value = convert_decimals<vectorized::DataTypeDecimal<vectorized::Decimal128>, 
+                                vectorized::DataTypeDecimal<vectorized::Decimal128>>(
+                                value, scale, 9);
+        }
+        column_data.emplace_back(value);
+    }
+    return Status::OK();
+}
+
+Status arrow_column_to_doris_column(const arrow::Array* arrow_column,
```
Comment (HappenLee) [inline be/src/vec/utils/arrow_column_to_doris_column.cpp:232]: Add UT for this class

File: be/src/vec/exec/vparquet_scanner.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+
+Status VParquetScanner::get_next(std::vector<MutableColumnPtr>& columns, bool* eof) {
+    // overall of type converting: 
+    // arrow type ==arrow_column_to_doris_column==> primitive type(PT0) ==cast_src_block==>
+    // primitive type(PT1) ==materialize_block==> dest primitive type
+    SCOPED_TIMER(_read_timer);
+    // init arrow batch
+    {
+        Status st = init_arrow_batch_if_necessary();
+        if (!st.ok()) {
```
Comment (HappenLee) [inline be/src/vec/exec/vparquet_scanner.cpp:137]: the code is weird? rethink the logic?

File: be/src/vec/exec/vparquet_scanner.cpp [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    RETURN_IF_ERROR(eval_conjunts(&src_block));
+    // materialize, src block => dest columns
+    RETURN_IF_ERROR(materialize_block(&src_block, columns));
+    *eof = _scanner_eof;
+    return Status::OK();
+}
+
+// eval conjuncts, for example: t1 > 1
+Status VParquetScanner::eval_conjunts(Block* block) {
+    for (auto& vctx : _pre_filter_vctxs) {
```
Comment (HappenLee) [inline be/src/vec/exec/vparquet_scanner.cpp:189]: may should change `_pre_filter_vctxs` to a tree to avoid the each time of copy operation

**Comment by yinzhijian** [comment]
close this pr, use this https://github.com/apache/incubator-doris/pull/9433

**Pull request closed without merging**